judge/views/user.py

```diff
@@ -507,3 +512,106 @@ def post(self, request, *args, **kwargs):
+        )
+        send_mail(
+            self.activate_subject_template_name, self.activate_email_template_name, context, from_email=None,
+            to_email=new_email, html_email_template_name=self.activate_html_email_template_name,
+        )
+
+        return generic_message(
+            self.request,
+            _('Email change requested'),
+            _('Please click on the link sent to the new email address.'),
+        )
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
+        return kwargs
+
+    def post(self, request, *args, **kwargs):
+        key = f'emailchange!{request.META["REMOTE_ADDR"]}'
+        cache.add(key, 0, timeout=settings.DMOJ_EMAIL_CHANGE_LIMIT_WINDOW)
+        if cache.incr(key) > settings.DMOJ_EMAIL_CHANGE_LIMIT_COUNT:
+            return HttpResponse(_('You have sent too many email change requests. Please try again later.'),
+                                content_type='text/plain', status=429)
+        return super().post(request, *args, **kwargs)
+
+
+class EmailChangeActivateView(LoginRequiredMixin, View):
+    class EmailChangeFailedError(Exception):
+        pass
+
+    def update_user_email(self, request, activation_key):
+        signer = signing.TimestampSigner()
+        try:
+            data = signer.unsign_object(
+                base64.urlsafe_b64decode(activation_key.encode()).decode(),
+                max_age=settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES * 60,
+            )
+        except (binascii.Error, signing.BadSignature):
+            raise self.EmailChangeFailedError(_('Invalid activation key. Please try again.'))
+        except signing.SignatureExpired:
+            raise self.EmailChangeFailedError(_('This request has expired. Please try again.'))
+        if data['id'] != request.user.id:
+            raise self.EmailChangeFailedError(
+                _('Please try again from the account this email change was originally requested from.'),
+            )
+        with revisions.create_revision(atomic=True):
+            if User.objects.filter(email=data['email']).exists():
+                raise self.EmailChangeFailedError(
+                    _('The email originally requested is in use. Please try again with a new email.'),
+                )
+            request.user.email = data['email']
+            request.user.save()
+            revisions.set_user(request.user)
+            revisions.set_comment(_('Changed email address'))
+
+    def get(self, request, *args, **kwargs):
+        try:
+            self.update_user_email(request, kwargs['activation_key'])
+        except self.EmailChangeFailedError as e:
+            return generic_message(request, _('Email change failed'), str(e), status=403)
+        else:
+            return generic_message(
+                request,
+                _('Email change succeeded'),
+                _('The email attached to your account has been changed.'),
```

Comment:
If it is easy (and I think we have the data here?) we should say to what in this message.